[android] Run tests for libdispatch in x64 emulator (wip)

weliveindetail opened the pull request:
Early preview for a ctest-based library that raises some questions:
* If we need additional scripts, where should we put them? For example `ctest_mock.sh` (we might be able to avoid this one).
* On my local machine I have to grant `adb` network access once. How to do that in CI?
* How to track the emulator process and make sure: (1) it booted, (2) it shut down

weliveindetail committed:
[android] Run tests for libdispatch in x64 emulator (wip)

weliveindetail commented:
Thanks for the feedback and links @andrurogerz. I am sure a GitHub workflow would be a good solution, but I think we want to run this in Swift CI. Thus, we need to implement it in the build script.

weliveindetail commented:
> If we need additional scripts, where should we put them?

libc++ has `utils/ci` for such scripts in upstream LLVM: https://github.com/llvm/llvm-project/tree/main/libcxx/utils/ci/vendor/android

> How to track the emulator process and make sure: (1) it booted, (2) it shut down

We likely find an answer in the above scripts

> On my local machine I have to grant adb network access once. How to do that in CI?

For the moment this question remains as it's Windows-specific. We will see with trial and error.

weliveindetail commented:
Thanks for your notes @finagolfin!

This is meant as a first test balloon to run something in the emulator at all in the Swift CI for Windows. I considered LIT-based tests more complicated and wanted to try something simple first. Happy to collaborate if there is overlap with Linux!

> You're probably aware that the compiler validation suite currently supports running its executable tests in the emulator through adb.

Yes, I literally [found them](https://github.com/thebrowsercompany/swift-build/issues/229#issuecomment-2602166992) 2 days ago :) That will help a lot with tests for Swift itself and out-of-tree libs like xctest!

> Andrew is right that you're better off using the existing github actions for that, if those are available in your Windows CI

I don't see how we would run these on ci.swift.org, is that possible?

weliveindetail commented:
@swift-ci Please test Windows

weliveindetail committed:
fixup! [android] Run tests for libdispatch in x64 emulator (wip)

weliveindetail committed:
[dev] Temporarily enable Android x86_64 in CI and test only dispatch

weliveindetail committed:
Properly spawn and shutdown both, adb and the emulator

weliveindetail commented:
Pushed a number of fixes and try to save some CI resources temporarily (only build Windows x64 + Android x86_64, only test dispatch).

@swift-ci Please test Windows

finagolfin commented:
> I don't see how we would run these on ci.swift.org, is that possible?

I don't know how the official Windows CI works. If it doesn't allow running github action workflows, then no, it's not possible.

I think Andrew's point is that you'd be better off running these tests in your existing github-based CI first, before proposing them for non-github-based CI that requires more configuration work.

weliveindetail committed:
Switch to cross-compiled ctest driver

weliveindetail commented:
In order to test Android in Swift CI, we need the `emulator` and `adb` from [Android cmdline-tools](https://developer.android.com/tools/releases/cmdline-tools). We can download and install them pretty quickly.

However, my above attempt failed, because the machine's Java Runtime version is outdated (55 is from 2018). @shahmishal Is there a chance that the build machines get an update? (61 is from 2021)
```
T:\android-cmdline-tools.zip not found. Downloading ...
Extracting 'android-cmdline-tools.zip' ...
Error: LinkageError occurred while loading main class com.android.sdklib.tool.sdkmanager.SdkManagerCli
	java.lang.UnsupportedClassVersionError: com/android/sdklib/tool/sdkmanager/SdkManagerCli has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0
Error: Error: sdkmanager.bat exited with code 1.
Invocation:
  T:\android-ndk-r26b\.temp\cmdline-tools\bin\sdkmanager.bat --sdk_root=T:\android-ndk-r26b cmdline-tools;latest --channel=3
```

finagolfin commented:
If you TBC devs don't control the Windows environment in that CI, it may be using [one of the Windows Docker files](https://github.com/swiftlang/swift-docker/tree/main/swift-ci/main/windows), in which case you can submit a Java update change there yourself, @weliveindetail.

weliveindetail commented:
Thanks @finagolfin! Yes, I'd like to consider that once I double-checked it's the right image source.

weliveindetail committed:
For the moment install Java Runtime on the fly

weliveindetail commented:
Unfortunately, the Windows builders haven't been ported to docker. In order to unblock progress we're installing Java on the fly now. This is a terrible workaround though. It's a waste of time and resources.

weliveindetail commented:
@swift-ci Please test Windows

weliveindetail commented:
The installation failed with:
```
T:\android-ndk-r26b-windows.zip not found. Downloading ...
Extracting 'android-ndk-r26b-windows.zip' ...
T:\android-cmdline-tools.zip not found. Downloading ...
T:\microsoft-jdk.zip not found. Downloading ...
Extracting 'microsoft-jdk.zip' ...
'android-cmdline-tools.zip' is already extracted and up to date.
Error: The term 'T:\android-ndk-r26b\.temp\cmdline-tools\bin\sdkmanager.bat' is not recognized
```

Must be an issue with `Extract-ZipFile`

weliveindetail committed:
Fix invalid 'zip is already extracted and up to date'

weliveindetail committed:
Add timer for FetchDependencies

weliveindetail committed:
utils: trim whitespace in build.ps1 (NFC)

weliveindetail commented:
@swift-ci Please test Windows

weliveindetail commented:
AVD failed to create a virtual device:
```
Error: Package path is not valid. Valid system image paths are:
null
Error: Error: avdmanager.bat exited with code 1.
Invocation:
  T:\android-ndk-r26b\cmdline-tools\latest\bin\avdmanager.bat create avd --name swift-test-device --package system-images;android-29;default;x86_64
```

weliveindetail committed:
Don't mix up release NDK with the one installed via cmdline-tools

weliveindetail commented:
The build failed before any of the changes in `$WebClient.DownloadFile($URL, $Destination)`:
```
T:\android-ndk-r26b-windows.zip not found. Downloading ...
Error: Exception calling "DownloadFile" with "2" argument(s): "The operation has timed out."
    at DownloadAndVerify, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 645
    at Fetch-Dependencies, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 795
```

Let's try that again.
@swift-ci Please test Windows

weliveindetail commented:
We reach that step now that creates the virtual device and need to figure out how to avoid the interactive config:
```
Auto-selecting single ABI x86_64
Do you wish to create a custom hardware profile? [no] offset 0, count -1, length 256
Error: Error: avdmanager.bat exited with code 1.
Invocation:
  T:\android-sdk\cmdline-tools\latest\bin\avdmanager.bat create avd --name "swift-test-device" --package "system-images;android-29;default;x86_64"
```

weliveindetail committed:
Disable interactive mode in avdmanager

weliveindetail commented:
@swift-ci Please test Windows

weliveindetail committed:
Fix emulator path, drop ndk package, add comments

weliveindetail committed:
Rename: NDKDir -> SDKDir

weliveindetail committed:
Revert "Switch to cross-compiled ctest driver"

This reverts commit 6ceb869500f28b0979f635debd82e4e38927f921.

weliveindetail committed:
Polish script-based driver

weliveindetail commented:
The build failed due to an unrelated issue in LLDB:
```
llvm-project\lldb\source\Plugins\LanguageRuntime\Swift\ReflectionContext.cpp(405): error C2039: 'ResumeAsyncContext': is not a member of 'swift::reflection::ReflectionContext<swift::External<swift::NoObjCInterop<swift::RuntimeTarget<8>>>>::AsyncTaskInfo'
```

weliveindetail committed:
Use proposed ctest-android.sh from libdispatch

weliveindetail commented:
Cross-compiling CTest for Android and using it as a test-driver doesn't seem like a good way forward. We need to build a different ctest binary for each target arch from source, post-process the config file (CTestTestfile.cmake) to transform absolute paths and still travers all executables manually to `chmod +x` them in the emulator. I keep the CMake build step here as a draft https://github.com/swiftlang/swift/pull/79013 (we might use it elsewhere), but revert this PR to the script-based approach.

https://github.com/swiftlang/swift-corelibs-libdispatch/pull/855 proposes such a script in libdispatch. The benefit of having it there is that CMake can fill in the targets at configuration time. This way we just need to push the binary directory and run the script.

Let's see if this will test the PRs together.
@swift-ci please test Windows

weliveindetail committed:
[dev] Add version dumps to see what runs and what hangs

weliveindetail committed:
[dev] Actually log device info before we run into wait-for-device for debugging

weliveindetail commented:
> Let's see if this will test the PRs together.

Nice, that worked! It checked out https://github.com/swiftlang/swift-corelibs-libdispatch/commit/215e85782e9208ddefaeee169c5af418cffc1f0c

We did reach the `Test-Dispatch` build step and failed with:
```
* daemon not running; starting now at tcp:5037
* daemon started successfully
Build timed out (after 60 minutes). Marking the build as aborted.
```

We likely hang in `wait-for-device`, let's run this early and dump device status for debugging.

https://github.com/swiftlang/swift-corelibs-libdispatch/pull/855
@swift-ci please test Windows

weliveindetail committed:
[dev] Avoid exit if get-state fails and dump even more logs

weliveindetail commented:
`adb get-state @swift-test-device` returned `error: no devices/emulators found` and bailed out. Fixed that and added more dumps of logs.

@swift-ci please test Windows

weliveindetail committed:
[dev] Temporarily drop stdout/stderr redirects from Start-Process

weliveindetail commented:
Emulator didn't start. Logs were not dumped. Let's try again without redirects.

@swift-ci please test Windows

weliveindetail committed:
[dev] Temporarily drop log dumps as well

finagolfin commented:
Wouldn't it be easier to develop this first locally _then_ check the resulting code on the CI? 😄

weliveindetail commented:
It works locally. I am trying to figure out that the issue is in CI.

weliveindetail commented:
@swift-ci please test Windows

xavgru12 commented:
Have you tried updating your local checkout(all dependency repositories)?

weliveindetail commented:
The issue here is that the Android emulator fails to start. There is no relation to checkouts of swift repos. All invocations of externals programs we do in build.ps1 so far are blocking. This is different here, we need the emulator to run detached. 

Locally, this works for me with this Powershell command:
```
$emulator = "$BinaryCache\android-sdk\emulator\emulator.exe"
Start-Process -FilePath $emulator -ArgumentList "@swift-test-device"
```

I can query the state of the emulated device (should return `device` status) and logs like this:
```
$adb = "$BinaryCache\android-sdk\platform-tools\adb.exe"
& $adb get-state "@swift-test-device"
& $adb logcat -d @swift-test-device
```

The fixed sleep time I inserted temporarily can be avoided and we just need this to wait until the device booted:
```
$adb "wait-for-device"
```

On the CI machine none of it works. The current output is:
```
T:\android-sdk\platform-tools\adb.exe get-state @swift-test-device
* daemon not running; starting now at tcp:5037
* daemon started successfully
error: no devices/emulators found
T:\android-sdk\platform-tools\adb.exe logcat -d @swift-test-device
- waiting for device -
```

 If you have ideas for this specific issue, please let me know.

weliveindetail committed:
Merge branch 'main' into android-ctest-libdispatch

weliveindetail committed:
[dev] Dump virtual devices with avdmanager to confirm that swift-test-device exists

weliveindetail commented:
https://github.com/swiftlang/swift-corelibs-libdispatch/pull/855
@swift-ci please test Windows

weliveindetail committed:
Set Java environment again when starting the emulator

weliveindetail commented:
Looks like we need the Java runtime for the emulator as well and not only for for cmdline-tools. This worked locally, because the Java Runtime on my host isn't outdated like the one on the CI machines. I will retry once the current run reached the timeout.

weliveindetail commented:
https://github.com/swiftlang/swift-corelibs-libdispatch/pull/855
@swift-ci please test Windows

weliveindetail commented:
Still:
```
error: no devices/emulators found
```

weliveindetail committed:
[dev] Run and teardown the emulator with debug output and multipe attempts

weliveindetail commented:
I can reproduce this locally by deleting the virtual device before running the build script. It seems that we cannot run a virtual device if we just created it, in particular, in the same process tree. The emulator process crashes without further notice. We can even wait and attempt it again and it crashes the same way. My latest patch demonstrates that. Running the build script again, however, works perfectly. So, this isn't an issue with permissions or execution environment.

Instead, I assume that the avdmanager keeps a file handle open and the emulator fails to lock it. Once we close the build script process, the handle is released and next time the emulator can lock it, because we don't need to recreate the virtual devices. I will double-check and see how to close them.

Anyway, let's run through the current state one more time.

https://github.com/swiftlang/swift-corelibs-libdispatch/pull/855
@swift-ci please test Windows

weliveindetail commented:
Build bot failed as expected:
```
ANDROID_SDK_HOME = T:\android-sdk
T:\android-sdk\cmdline-tools\latest\bin\avdmanager.bat list avd reports:
Available Android Virtual Devices:
Create Android virtual device for arch x86_64
T:\android-sdk\cmdline-tools\latest\bin\avdmanager.bat list avd reports:
Available Android Virtual Devices:
    Name: swift-test-device-x86_64
    Path: T:\android-sdk\.android\avd\swift-test-device-x86_64.avd
  Target: Default Android System Image
          Based on: Android 10.0 ("Q") Tag/ABI: default/x86_64
  Sdcard: 512 MB
T:\android-sdk\emulator\emulator.exe -list-avds reports:
swift-test-device-x86_64
Start Android emulator for arch x86_64
Waiting for process 4584 to start
Process 4584 failed to start, trying again...
Waiting for process 9360 to start
Process 9360 failed to start, trying again...
Waiting for process 9668 to start
Process 9668 failed to start, trying again...
Waiting for process 5356 to start
Process 5356 failed to start, trying again...
Waiting for process 7624 to start
Error: Android emulator process 7624 failed to start.
```

weliveindetail committed:
Running avdmanager via Start-Process fixes the emulator crashes

weliveindetail commented:
Trial and error found a fix locally. Let's see if CI reproduces that 🤞
@swift-ci please test Windows

weliveindetail commented on utils/build.ps1:
```diff
@@ -916,6 +971,104 @@ function Fetch-Dependencies {
+
+      Write-Host "Waiting while Android emulator boots in process $AndroidEmulatorPid"
+      $adb = "$BinaryCache\android-sdk\platform-tools\adb.exe"
+      Invoke-Program $adb "wait-for-device"
```
No the build bot hangs here in `adb wait-for-device`

weliveindetail reviewed (commented):
We got past the emulator startup 🙌

weliveindetail committed:
Let's grant adb a few more attempts

weliveindetail commented:
@swift-ci please test Windows

weliveindetail commented:
No luck with that
```
T:\android-sdk\emulator\emulator.exe -list-avds reports:
swift-test-device-x86_64
Start Android emulator for arch x86_64
Waiting for process 6632 to start
Waiting while Android emulator boots in process 6632
adb wait-for-device running in process 7648
* daemon not running; starting now at tcp:5037
* daemon started successfully
adb failed to connect. Shutting it down.
adb wait-for-device running in process 3928
adb failed to connect. Shutting it down.
adb wait-for-device running in process 6132
adb failed to connect. Shutting it down.
adb wait-for-device running in process 10728
adb failed to connect. Shutting it down.
adb wait-for-device running in process 6324
Error: Android Debug Bridge process 6324 failed to connect.
    at <ScriptBlock>, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 1047
    at Isolate-EnvVars, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 610
    at AndroidEmulator-Run, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 993
    at <ScriptBlock>, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 3214
  From System.Management.Automation.RuntimeException: Android Debug Bridge process 6324 failed to connect.
error: no emulator detected
```

finagolfin commented:
Are you familiar with [the popular ReactiveCircus github workflow to start an Android emulator on macOS/linux](https://github.com/ReactiveCircus/android-emulator-runner)? They have it create some kind of polling loop in Typescript that waits to make sure the emulator starts up. Maybe you could look at their OSS code and find something useful for your Windows emulator startup here.

weliveindetail commented:
No I wasn't aware of it. I am afraid the issues here are very Windows specific, but it's still a good source for common things to try. Like emulator flags `-no-snapshot-save -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none` for example. Thanks

weliveindetail committed:
Try to grant network access permission to adb

weliveindetail committed:
Merge branch 'main' into android-ctest-libdispatch

weliveindetail commented:
@swift-ci please test Windows

weliveindetail committed:
[dev] More emulator params and dump verbose outputs

weliveindetail commented:
@swift-ci please test Windows

weliveindetail committed:
[dev] Run -accel-check and dump the output

weliveindetail committed:
Dump logs in erro case as well

weliveindetail commented:
@swift-ci please test Windows
